modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/Pipe.java

```diff
@@ -245,17 +246,27 @@ public int produce(final ContentDecoder decoder) throws IOException {
         lock.lock();
         try {
             setInputMode(buffer);
+            int totalRead = 0;
             int bytesRead;
             try {
-                bytesRead = decoder.read(buffer.getByteBuffer());
+                // Drain the decoder until the end of the underlying stream is found or until the Pipe#buffer is full.
+                // bytesRead = -1 means reached out to the end of underlying stream.
+                // bytesRead = 0 means Pipe's input buffer is full.
+                while ((bytesRead = decoder.read(buffer.getByteBuffer())) > 0) {
+                    totalRead += bytesRead;
+                }
             } catch (TruncatedChunkException ignore) {
```

Comment:
since here we are handling the TruncatedChunkException, let's rename the variable to ex (it's not ignored).

Reply:
Fixed in 1a225662c71267f47edbbee23cb3dcf7c2106fd3